### PR TITLE
Front authentication

### PR DESCRIPTION
**Pull request by LukaszBielsko**

### CODE REVIEW

File: frontend/src/components/AuthHandler.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { useHistory } from "react-router-dom";
+import { useQuery } from "../utils";
+
+const AuthHandler = ({ setIsLoggedIn }) => {
+  const query = useQuery();
+  const token = query.get("auth_token");
+  window.localStorage.setItem("til_token", token);
+  if (token) setIsLoggedIn(true);
+  const history = useHistory();
+  history.push("/");
```
Comment (MarcinCholewka) [inline frontend/src/components/AuthHandler.js]:
```suggestion
  const query = useQuery();
  const token = query.get("auth_token");
  const history = useHistory();
  
  window.localStorage.setItem("til_token", token);
  
  if (token) setIsLoggedIn(true);
  
  history.push("/");
```

File: frontend/src/App.js [hunk 2]
```diff
@@ -1,33 +1,31 @@
...
+import NonAuthenticatedApp from "./nonAuthenticated";
 
 function App() {
+  const [isLoggedIn, setIsLoggedIn] = useState(false);
+
+  useEffect(() => {
+    setIsLoggedIn(isAuthenticated());
+  }, []);
+
+  const renderApp = isLoggedIn ? (
```
Comment (MarcinCholewka) [inline frontend/src/App.js]: why use hook `isLoggedIn` as the same output you have with 'isAuthenticated' and that is what you're checking here
```suggestion
  const renderApp = IsAuthenticated ? (
```

File: frontend/src/App.js [hunk 3]
```diff
@@ -1,33 +1,31 @@
...
+import AuthenticatedApp from "./authenticated";
+import { isAuthenticated } from "./utils";
+import NonAuthenticatedApp from "./nonAuthenticated";
 
 function App() {
+  const [isLoggedIn, setIsLoggedIn] = useState(false);
+
+  useEffect(() => {
+    setIsLoggedIn(isAuthenticated());
+  }, []);
```
Comment (MarcinCholewka) [inline frontend/src/App.js]: not needed, use `IsAuthenticated` to render the correct component ✂️

File: frontend/src/components/MainRoutes.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+      <Route path="/random-post">
+        <RandomPost />
+      </Route>
+      <Route path="/stats">
+        <Stats />
+      </Route>
+      <Route path="/">
+        <Posts />
+      </Route>
+    </Switch>
```
Comment (MarcinCholewka) [inline frontend/src/components/MainRoutes.js]:
```suggestion
    <Switch>
      <Route exact path="/">
        <Posts />
      </Route>
      <Route path="/random-post">
        <RandomPost />
      </Route>
      <Route path="/stats">
        <Stats />
      </Route>
    </Switch>
```

File: frontend/src/components/Posts.js [hunk 5]
```diff
@@ -10,24 +10,26 @@ const Posts = props => {
   }, []);
 
   const fetchPosts = async () => {
-    const response = await fetch(
-      "http://localhost:5000/posts?_sort=id&_order=DESC"
-    ).catch(error => console.error("Error:", error));
+    const response = await fetch("/api/posts").catch(error =>
+      console.error("Error:", error)
+    );
```
Comment (MarcinCholewka) [inline frontend/src/components/Posts.js]:
```suggestion
    const response = await fetch("/api/posts")
      .catch(error =>
        console.error("Error:", error)
    );
```

File: frontend/src/components/Posts.js [hunk 6]
```diff
@@ -10,24 +10,26 @@ const Posts = props => {
...
 
     if (!response) {
       return;
     }
-
     const fetchedPosts = await response.json();
 
-    setPosts(fetchedPosts);
+    // setPosts(fetchedPosts);
```
Comment (MarcinCholewka) [inline frontend/src/components/Posts.js]: Is this necessary? 🤔

File: frontend/src/components/Posts.js [hunk 7]
```diff
@@ -10,24 +10,26 @@ const Posts = props => {
...
 
   return (
     <div className="posts">
-      {posts.map(postObject => (
-        <Post key={postObject.id} {...postObject} />
-      ))}
+      <>
+        <p>posts listed here</p>
+        {posts.map(postObject => (
+          <Post key={postObject.id} {...postObject} />
```
Comment (MarcinCholewka) [inline frontend/src/components/Posts.js]: singular post will be enough
```suggestion
        {posts.map(post => (
          <Post key={post.id} {...post} />
```

File: frontend/src/components/RandomPost.js [hunk 8]
```diff
@@ -0,0 +1,8 @@
+import React from "react";
+
+const RandomPost = props => {
+  //fetch random post
```
Comment (MarcinCholewka) [inline frontend/src/components/RandomPost.js]: Is this necessary? 🤔

File: frontend/src/nonAuthenticated/index.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import React from "react";
+import MainRoutes from "../components/MainRoutes";
+import SideBar from "../components/SideBar";
+
+const NonAuthenticatedApp = props => {
```
Comment (MarcinCholewka) [inline frontend/src/nonAuthenticated/index.js]: unused props - remove them
```suggestion
const NonAuthenticatedApp =() => {
```

File: frontend/src/utils.js [hunk 10]
```diff
@@ -10,3 +12,18 @@ export const postData = async (url, data) => {
 
   return response.ok;
 };
+
+export const useQuery = () => {
+  const queryString = useLocation().search;
```
Comment (MarcinCholewka) [inline frontend/src/utils.js]: add empty line 😱

File: frontend/src/utils.js [hunk 11]
```diff
@@ -10,3 +12,18 @@ export const postData = async (url, data) => {
...
+  return new URLSearchParams(queryString);
+};
+
+export const isAuthenticated = () => {
+  const token = window.localStorage.getItem("til_token");
```
Comment (MarcinCholewka) [inline frontend/src/utils.js]: add empty line 😱

File: frontend/src/utils.js [hunk 12]
```diff
@@ -10,3 +12,18 @@ export const postData = async (url, data) => {
...
+  if (token) return true;
+  else return false;
```
Comment (MarcinCholewka) [inline frontend/src/utils.js]: `else` not needed
```suggestion
  if (token) return true;
  
  return false;
```

File: backend/config/config.exs [hunk 13]
```diff
@@ -1,4 +1,4 @@
-# This file is responsible for configuring your application
+e# This file is responsible for configuring your application
```
[the remark this answers is not in the document]
Comment (LukaszBielsko) [reply backend/config/config.exs]: nope
deleted

File: frontend/src/App.css [hunk 14]
```diff
@@ -1,25 +1,17 @@
-.add-post {
-  margin: 20px auto;
-  display: block;
+* {
+  padding: 5px;
+  /* margin: 10px; */
```
[the remark this answers is not in the document]
Comment (LukaszBielsko) [reply frontend/src/App.css]: deleted

File: frontend/src/App.css [hunk 15]
```diff
@@ -1,25 +1,17 @@
...
-  display: flex;
-  justify-content: center;
-}
-
-nav li {
-  margin: 0 20px;
+.side-nav-bar-list {
   list-style: none;
+  display: flex;
+  flex-direction: column;
```
[the remark this answers is not in the document]
Comment (LukaszBielsko) [reply frontend/src/App.css]: sorted ascendingly

File: frontend/src/App.js [hunk 16]
```diff
@@ -1,33 +1,31 @@
-import React from "react";
-import { BrowserRouter as Router, Switch, Route, Link } from "react-router-dom";
+import React, { useState, useEffect } from "react";
+import { BrowserRouter as Router, Route } from "react-router-dom";
 import "./App.css";
```
[the remark this answers is not in the document]
Comment (LukaszBielsko) [reply frontend/src/App.js]: done

File: frontend/src/components/AuthHandler.js (hunk 1 above)
Comment (LukaszBielsko) [reply frontend/src/components/AuthHandler.js]: changed to proposed

File: frontend/src/components/MainRoutes.js (hunk 4 above)
Comment (LukaszBielsko) [reply frontend/src/components/MainRoutes.js]: changed to proposed

File: frontend/src/components/Posts.js (hunk 5 above)
Comment (LukaszBielsko) [reply frontend/src/components/Posts.js]: changed to proposed

File: frontend/src/components/Posts.js (hunk 7 above)
Comment (LukaszBielsko) [reply frontend/src/components/Posts.js]: changed to proposed

File: frontend/src/components/Posts.js (hunk 6 above)
Comment (LukaszBielsko) [reply frontend/src/components/Posts.js]: this feature will be implemented later 
added TODO as discussed

File: frontend/src/components/Posts.js (hunk 7 above)
Comment (LukaszBielsko) [reply frontend/src/components/Posts.js]: good idea
changed to proposed

File: frontend/src/components/RandomPost.js (hunk 8 above)
Comment (LukaszBielsko) [reply frontend/src/components/RandomPost.js]: this feature will be implemented later
added TODO as discussed

File: frontend/src/nonAuthenticated/index.js (hunk 9 above)
Comment (LukaszBielsko) [reply frontend/src/nonAuthenticated/index.js]: deleted as requested

File: frontend/src/utils.js (hunk 10 above)
Comment (LukaszBielsko) [reply frontend/src/utils.js]: done

File: frontend/src/utils.js (hunk 11 above)
Comment (LukaszBielsko) [reply frontend/src/utils.js]: added empty line

File: frontend/src/utils.js (hunk 12 above)
Comment (LukaszBielsko) [reply frontend/src/utils.js]: removed else statement and changed to proposed

File: frontend/src/App.js (hunk 2 above)
Comment (LukaszBielsko) [reply frontend/src/App.js]: I use setIsLoggedIn to set isLoggedIn in components further down in the tree
if I use only isAuthenticated and delete token in the component down in the tree isAuthenticated will not be called and app will not re-render 

if there is another solution I will implement it

File: frontend/src/App.js (hunk 3 above)
Comment (LukaszBielsko) [reply frontend/src/App.js]: I use setIsLoggedIn to set isLoggedIn in components further down in the tree
if I use only isAuthenticated and delete token in the component down in the tree isAuthenticated will not be called and app will not re-render 

if there is another solution I will implement it

File: frontend/src/App.js (hunk 2 above)
Comment (MarcinCholewka) [reply frontend/src/App.js]: 👍

File: frontend/src/App.css [hunk 17]
```diff
@@ -1,24 +1,15 @@
...
-  margin: 20px auto;
-  display: block;
+* {
+  padding: 5px;
 }
 
-.App {
-  margin: 20px;
+.side-nav-bar {
+  border: 1px solid black;
```
Comment (MarcinCholewka) [inline frontend/src/App.css]: try to use hex values for colors.
I recommend the website [Flat UI Color](https://flatuicolors.com/)
Black is `#000000`

Comment (LukaszBielsko) [reply frontend/src/App.css]: changed

**Pull request merged**